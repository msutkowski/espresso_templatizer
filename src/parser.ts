import type { HTMLElement, Node } from "node-html-parser";
import { NodeType } from "node-html-parser";

type GleamableElement = {
  type: "element" | "text";
  tag_name: string;
  attributes: { name: string; value: unknown }[];
  children: Array<GleamableElement | GleamableText>;
};

type GleamableText = {
  type: "text";
  content: string;
};

type GleamableNode = GleamableElement | GleamableText | null;

export const isHTMLElement = (node: Node): node is HTMLElement =>
  node.nodeType === NodeType.ELEMENT_NODE;

export const isTextNode = (node: Node): node is HTMLElement =>
  node.nodeType === NodeType.TEXT_NODE;

export const isGleamableElement = (
  node: GleamableNode
): node is GleamableElement => node?.type === "element";

export const isGleamableText = (node: GleamableNode): node is GleamableText =>
  node?.type === "text";

export function traverseDOM(node: HTMLElement | Node): GleamableNode {
  if (isTextNode(node)) {
    return {
      type: "text",
      content: node.textContent.trim(),
    } as GleamableText;
  } else if (isHTMLElement(node)) {
    const element: GleamableNode = {
      type: "element",
      tag_name: node.tagName?.toLowerCase() || "div",
      attributes: getAttributes(node),
      children: [],
    };

    // Loop over child nodes
    for (let i = 0; i < node.childNodes.length; i++) {
      const childNode = node.childNodes[i];

      // Recursively call the function for child nodes
      const childElement = traverseDOM(childNode);

      // Add the child Element object to the children array
      if (childElement) {
        element.children.push(childElement);
      }
    }

    return element;
  } else {
    // Other node types (e.g., comment nodes, etc.)
    return null;
  }
}

function getAttributes(node: HTMLElement) {
  return Object.entries(node.attributes).map(([key, value]) => ({
    name: key,
    value,
  }));
}

export function toGleam(obj: GleamableNode) {
  if (isGleamableElement(obj)) {
    let code = `t("${obj.tag_name}")`;

    if (obj.attributes.length > 0) {
      for (const attr of obj.attributes) {
        code += ` |> a("${attr.name}", "${attr.value}")`;
      }
    }

    const children = obj.children.map((child) => toGleam(child));
    if (children.length > 0) {
      code += ` |> c([${children.filter((f) => f !== 'txt("")').join(", ")}])`;
    }

    return code;
  } else if (isGleamableText(obj)) {
    return `txt("${obj.content}")`;
  } else {
    return "";
  }
}
