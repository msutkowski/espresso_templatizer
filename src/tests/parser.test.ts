import { describe, expect, test } from "vitest";
import parse from "node-html-parser";
import { toGleam, traverseDOM } from "../parser";
import { simpleTestHtml } from "./common";

describe("traverseDOM", () => {
  test("generates an internal representation for gleam node output", () => {
    const parsedDom = parse(simpleTestHtml);
    expect(traverseDOM(parsedDom)).toMatchSnapshot();
  });
});

describe("toGleam", () => {
  test("converts GleamableNodes into gleam code", () => {
    const gleamableNodes = traverseDOM(parse(simpleTestHtml));
    expect(toGleam(gleamableNodes)).toMatchSnapshot();
  });
});
