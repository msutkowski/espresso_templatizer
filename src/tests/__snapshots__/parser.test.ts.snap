// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toGleam > converts GleamableNodes into gleam code 1`] = `"t(\\"div\\") |> c([t(\\"span\\") |> c([t(\\"div\\") |> a(\\"class\\", \\"text-black\\") |> c([txt(\\"Text here\\")])])])"`;

exports[`traverseDOM > generates an internal representation for gleam node output 1`] = `
{
  "attributes": [],
  "children": [
    {
      "content": "",
      "type": "text",
    },
    {
      "attributes": [],
      "children": [
        {
          "content": "",
          "type": "text",
        },
        {
          "attributes": [
            {
              "name": "class",
              "value": "text-black",
            },
          ],
          "children": [
            {
              "content": "Text here",
              "type": "text",
            },
          ],
          "tag_name": "div",
          "type": "element",
        },
        {
          "content": "",
          "type": "text",
        },
      ],
      "tag_name": "span",
      "type": "element",
    },
    {
      "content": "",
      "type": "text",
    },
  ],
  "tag_name": "div",
  "type": "element",
}
`;
